pkg/healthcheck/healthcheck_multicluster.go

```diff
@@ -491,3 +629,300 @@ func (hc *HealthChecker) checkRemoteClusterGatewaysHealth(ctx context.Context) e
+		if err != nil {
+			return err
+		}
+
+		for _, svc := range mirrorServices.Items {
+			// Check if there is a relevant end-point
+			endpoint, err := hc.kubeAPI.CoreV1().Endpoints(svc.Namespace).Get(svc.Name, metav1.GetOptions{})
+			if err != nil || len(endpoint.Subsets) == 0 {
+				servicesWithNoEndpoints = append(servicesWithNoEndpoints, fmt.Sprintf("%s.%s of cluster: [%s], gateway: [%s/%s]", svc.Name, svc.Namespace, svc.Labels[k8s.RemoteClusterNameLabel], svc.Labels[k8s.RemoteGatewayNsLabel], svc.Labels[k8s.RemoteGatewayNameLabel]))
+			}
+		}
+
+		if len(servicesWithNoEndpoints) > 0 {
+			return fmt.Errorf("Some mirror services do not have endpoints:\n\t%s", strings.Join(servicesWithNoEndpoints, "\n\t"))
+		}
+		return nil
+	}
+	return &SkipError{Reason: "not checking muticluster"}
+}
+
+func (hc *HealthChecker) checkIfGatewayMirrorsHaveEndpoints() error {
+	if hc.Options.SourceCluster {
+
+		var gatewayMirrorsWithNoEndpoints []string
+		gatewayServices, err := hc.kubeAPI.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{LabelSelector: k8s.MirroredGatewayLabel})
+		if err != nil {
+			return err
+		}
+
+		for _, svc := range gatewayServices.Items {
+			// Check if there is a relevant end-point
+			endpoints, err := hc.kubeAPI.CoreV1().Endpoints(svc.Namespace).Get(svc.Name, metav1.GetOptions{})
+			if err != nil || len(endpoints.Subsets) == 0 {
+				gatewayMirrorsWithNoEndpoints = append(gatewayMirrorsWithNoEndpoints, fmt.Sprintf("%s.%s of cluster: [%s]", svc.Name, svc.Namespace, svc.Labels[k8s.RemoteClusterNameLabel]))
+			}
+		}
+
+		if len(gatewayMirrorsWithNoEndpoints) > 0 {
+			return fmt.Errorf("Some mirror services do not have endpoints:\n\t%s", strings.Join(gatewayMirrorsWithNoEndpoints, "\n\t"))
+		}
+		return nil
+	}
+	return &SkipError{Reason: "not checking muticluster"}
+}
+
+func (hc *HealthChecker) checkIfAllMirrorServicesHaveGatewayMirrors() error {
+
+	if hc.Options.SourceCluster {
+
+		var nonExistentGatewayMirrors []string
+		selector := fmt.Sprintf("%s,!%s", k8s.MirroredResourceLabel, k8s.MirroredGatewayLabel)
+		mirrorServices, err := hc.kubeAPI.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			return err
+		}
+
+		for _, svc := range mirrorServices.Items {
+			// Check if the relevant gateway mirror exists
+			gatewayMirrorName := svc.Labels[k8s.RemoteGatewayNameLabel] + "-" + svc.Labels[k8s.RemoteClusterNameLabel]
+			_, err := hc.kubeAPI.CoreV1().Services(svc.Labels[k8s.RemoteGatewayNsLabel]).Get(gatewayMirrorName, metav1.GetOptions{})
+			if err != nil {
+				nonExistentGatewayMirrors = append(nonExistentGatewayMirrors, fmt.Sprintf("%s.%s", gatewayMirrorName, svc.Annotations[k8s.RemoteGatewayNsLabel]))
+			}
+		}
+
+		if len(nonExistentGatewayMirrors) > 0 {
+			return fmt.Errorf("Some mirror services do not have a valid gateway mirror:\n\t%s", strings.Join(nonExistentGatewayMirrors, "\n\t"))
```

Comment:
This sounds like it will list the mirror services which don't have valid gateway mirrors but instead it lists the nonExistentGatewayMirrors.  I think listing the problematic mirror services is more useful.